### PR TITLE
feat: add retries for enqueuing graphile jobs

### PR DESCRIPTION
**Pull request by yakkomajuri**
## Problem

We're losing jobs when Aurora has small periods of unavailability due to scaling.

https://sentry.io/organizations/posthog2/issues/3477565970/?query=is%3Aunresolved+no+jobqueue&statsPeriod=14d

## Changes

- Retry enqueueing jobs into Graphile using an existing tested, traced, and instrumented retries mechanism 
- Refactors the mechanism referenced above to allow it to be used more generally for retries everywhere

## How did you test this code?

- Updated the tests for the retries mechanism
- Added tests for the JobQueueManager
- Tested manually

## Not in scope

- Tests for untouched bits of jobs functionality that are currently missing

### CODE REVIEW

**Comment by yakkomajuri** [comment]
I know what's up with the test errors, will address tomorrow. Just some tests that weren't updated

**Comment by mariusandra** [comment]
This is a disappointment with regards to Aurora. I remember we chose it just because it's a scalable cloud database that could scale up if we get a sudden spike of jobs. 

If instead it's a scalable database, except when it scales (probably as we're coming into a spike), then what's the point 🤦

**Comment by yakkomajuri** [comment]
Come to think about it - I wonder if part of the problem is that the graphile-worker lib internally is trying to reuse a connection and doesn't even try to establish a new one after a disconnect?

File: plugin-server/src/main/job-queues/job-queue-manager.ts [hunk 1]
```diff
@@ -50,7 +58,35 @@ export class JobQueueManager implements JobQueue {
...
 
-    async enqueue(jobName: string, job: EnqueuedJob): Promise<void> {
+    async enqueue(jobName: string, job: EnqueuedJob, instrumentationContext?: InstrumentationContext): Promise<void> {
+        const jobType = 'type' in job ? job.type : 'buffer'
+        const jobPayload = 'payload' in job ? job.payload : job.eventPayload
+        const pluginServerMode = this.pluginsServer.PLUGIN_SERVER_MODE ?? 'full'
+        await instrument(
+            this.pluginsServer.statsd,
+            {
+                metricName: jobName === JobName.PLUGIN_JOB ? 'vm.enqueuePluginJob' : 'vm.enqueueBufferJob',
```
Comment (macobo) [inline plugin-server/src/main/job-queues/job-queue-manager.ts:68]: So this is confusing.
- It makes an implicit assertion that there's only two different types of jobs. Which might be true for now, but will it be long-term?
- It prefixes the metric with `vm.` which just is a lie

Suggestion: 

```suggestion
                metricName: `job_queues_enqueue`
```

And add jobName as a tag instead.

File: plugin-server/src/main/job-queues/job-queue-manager.ts [hunk 2]
```diff
@@ -50,7 +58,35 @@ export class JobQueueManager implements JobQueue {
...
+                key: instrumentationContext?.key ?? '?',
+                tag: instrumentationContext?.tag ?? '?',
+                tags: { pluginServerMode, type: jobType },
```
Comment (macobo) [inline plugin-server/src/main/job-queues/job-queue-manager.ts:71]: There's no reason to add pluginServerMode tag - https://github.com/PostHog/posthog/blob/master/plugin-server/src/utils/db/hub.ts#L82-L84

File: plugin-server/src/main/job-queues/job-queue-manager.ts [hunk 3]
```diff
@@ -50,7 +58,35 @@ export class JobQueueManager implements JobQueue {
...
+            () =>
+                runRetriableFunction({
+                    hub: this.pluginsServer,
+                    metricPrefix: 'enqueueJob',
+                    metricName: jobName,
+                    metricTags: {
+                        pluginServerMode,
+                    },
+                    tryFn: async () => this._enqueue(jobName, job),
+                    catchFn: () => status.error('🔴', 'Exhausted attempts to enqueue job.'),
```
Comment (macobo) [inline plugin-server/src/main/job-queues/job-queue-manager.ts:83]: Problems:
- The log message isn't actionable or even clear. Suggestion: `Exhausted attempts to enqueue job, job was dropped.`
- We should log the `jobName, job` objects as well

Q:
- This is a `catchFn` not `finallyFn`? Does this swallow all errors or will sentry receive an error as well? T

File: plugin-server/src/utils/retries.ts [hunk 4]
```diff
@@ -0,0 +1,112 @@
+import { RetryError } from '@posthog/plugin-scaffold'
```
Comment (macobo) [inline plugin-server/src/utils/retries.ts:1]: The changes here are impossible to follow because we both moved a file and then refactored it in the same PR.

Please keep a clear commit history in cases like this or e.g. do the move in a follow-up PR as otherwise you make the reviewers job worse.

File: plugin-server/src/utils/retries.ts [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+            } catch (error) {
+                if (error instanceof RetryError) {
+                    error._attempt = attempt
+                    error._maxAttempts = maxAttempts
+                }
+                if (error instanceof RetryError && attempt < maxAttempts) {
+                    const nextRetryMs = getNextRetryMs(retryBaseMs, retryMultiplier, attempt)
+                    hub.statsd?.increment(`${metricPrefix}.${metricName}.RETRY`, {
+                        ...metricTags,
+                        attempt: attempt.toString(),
```
Comment (macobo) [inline plugin-server/src/utils/retries.ts:74]: We should be really conservative with tags. What value does attempt tag give us?

File: plugin-server/src/utils/retries.ts [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+            try {
+                await tryFn()
+            } catch (error) {
+                if (error instanceof RetryError) {
+                    error._attempt = attempt
+                    error._maxAttempts = maxAttempts
+                }
+                if (error instanceof RetryError && attempt < maxAttempts) {
+                    const nextRetryMs = getNextRetryMs(retryBaseMs, retryMultiplier, attempt)
+                    hub.statsd?.increment(`${metricPrefix}.${metricName}.RETRY`, {
```
Comment (macobo) [inline plugin-server/src/utils/retries.ts:72]: Note: Reverse-engineering actual metrics from code in cases of interpolation like this is hell. Do we actually need this? Could we just use `metricName` and kill the `metricPrefix`?

File: plugin-server/src/utils/retries.ts [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+        retryMultiplier = 2,
+    } = retriableFunctionPayload
+    return runInTransaction(
+        {
+            name: 'retryLoop',
+            op: metricName,
+            description: metricTags.plugin || '?',
+            data: {
+                metricName,
+                payload,
```
Comment (macobo) [inline plugin-server/src/utils/retries.ts:57]: Does nesting objects like this work nicely in sentry?

File: plugin-server/tests/job-queue-manager.test.ts [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    beforeEach(() => {
+        jest.clearAllMocks()
+        jobQueueManager = new JobQueueManager(mockHub)
+    })
+
+    describe('enqueue()', () => {
+        it('calls runRetriableFunction with the correct parameters', async () => {
+            await jobQueueManager.enqueue(JobName.PLUGIN_JOB, { type: 'foo' } as EnqueuedJob)
+            expect(runRetriableFunction).toHaveBeenCalled()
+            const runRetriableFunctionArgs = jest.mocked(runRetriableFunction).mock.calls[0][0]
```
Comment (macobo) [inline plugin-server/tests/job-queue-manager.test.ts:28]: Checking calls this way is a code smell.

Do you mean to do something like this:

```
expect(runRetriableFunction).toHaveBeenCalledWith({
  metricPrefix: 'enqueueJob',
  //...
})
```
Note you can use `expect.any()` and equivalent within the object if there's problematic arguments.

File: plugin-server/tests/job-queue-manager.test.ts [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+            expect(runRetriableFunctionArgs.metricTags).toEqual({ pluginServerMode: 'full' })
+            expect(runRetriableFunctionArgs.tryFn).not.toBeUndefined()
+            expect(runRetriableFunctionArgs.catchFn).not.toBeUndefined()
+            expect(runRetriableFunctionArgs.finallyFn).toBeUndefined()
+        })
+    })
+
+    describe('_enqueue()', () => {
+        it('enqueues jobs to the first available job queue', async () => {
+            jobQueueManager.jobQueues[0].enqueue = jest.fn()
```
Comment (macobo) [inline plugin-server/tests/job-queue-manager.test.ts:42]: Code nit:

Use `jest.spyOn()`

File: plugin-server/tests/job-queue-manager.test.ts [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+const mockHub: Hub = {
+    instanceId: new UUID('F8B2F832-6639-4596-ABFC-F9664BC88E84'),
+    promiseManager: new PromiseManager({ MAX_PENDING_PROMISES_PER_WORKER: 1 } as any),
+    JOB_QUEUES: 'fs',
+} as Hub
+
+describe('JobQueueManager', () => {
+    let jobQueueManager: JobQueueManager
+    beforeEach(() => {
+        jest.clearAllMocks()
```
Comment (macobo) [inline plugin-server/tests/job-queue-manager.test.ts:20]: This is not needed.

File: plugin-server/tests/job-queue-manager.test.ts [hunk 11]
```diff
@@ -0,0 +1,70 @@
+import { RetryError } from '@posthog/plugin-scaffold'
+
+import { runRetriableFunction } from '../src/utils/retries'
+import { UUID } from '../src/utils/utils'
+import { PromiseManager } from '../src/worker/vm/promise-manager'
+import { JobQueueManager } from './../src/main/job-queues/job-queue-manager'
+import { EnqueuedJob, Hub, JobName } from './../src/types'
+
+jest.mock('../src/utils/retries')
```
Comment (macobo) [inline plugin-server/tests/job-queue-manager.test.ts:9]: Does this test log? If so, please also mock /status util.

File: plugin-server/tests/job-queue-manager.test.ts [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+} as Hub
+
+describe('JobQueueManager', () => {
+    let jobQueueManager: JobQueueManager
+    beforeEach(() => {
+        jest.clearAllMocks()
+        jobQueueManager = new JobQueueManager(mockHub)
+    })
+
+    describe('enqueue()', () => {
```
Comment (macobo) [inline plugin-server/tests/job-queue-manager.test.ts:24]: Missing a test for error handling.

File: plugin-server/tests/worker/vm.test.ts [hunk 13]
```diff
@@ -1272,23 +1283,28 @@ describe('vm tests', () => {
             }
             await delay(1010)
 
-            expect(fetch).toHaveBeenCalledTimes(15)
+            expect(fetch).toHaveBeenCalledTimes(20)
```
Comment (macobo) [inline plugin-server/tests/worker/vm.test.ts:1286]: What is this change caused by?

File: plugin-server/src/main/job-queues/job-queue-manager.ts (hunk 1 above)
Comment (yakkomajuri) [reply plugin-server/src/main/job-queues/job-queue-manager.ts:68]: I was making sure we kept the metric that was already established before, but can get rid of that in favor of this new metric

File: plugin-server/src/utils/retries.ts (hunk 5 above)
Comment (yakkomajuri) [reply plugin-server/src/utils/retries.ts:74]: I believe you set this up in the past, happy to remove now

File: plugin-server/src/utils/retries.ts (hunk 6 above)
Comment (yakkomajuri) [reply plugin-server/src/utils/retries.ts:72]: Understand it was hard to parse the new changes from old, but this is effectively following what was already there.

File: plugin-server/src/utils/retries.ts (hunk 7 above)
Comment (yakkomajuri) [reply plugin-server/src/utils/retries.ts:57]: Unsure - again, this was here before, just it was named "event"

File: plugin-server/tests/worker/vm.test.ts (hunk 13 above)
Comment (yakkomajuri) [reply plugin-server/tests/worker/vm.test.ts:1286]: We stopped checking type errors on test files, so I noticed this file had a bunch and fixed them. One of them was that I added the missing `uuid` to all events. That causes the payload to be larger, making this test for buffering based on bytes flush more often. 

Also took me a while to figure this out, was very confused, super annoying.

File: plugin-server/tests/job-queue-manager.test.ts (hunk 8 above)
Comment (macobo) [reply plugin-server/tests/job-queue-manager.test.ts:28]: I have had zero problems and we use both `expect.any()`, `expect.objectContaining()` extensively. toMatchObject much less so though.

File: plugin-server/tests/job-queue-manager.test.ts (hunk 12 above)
Comment (yakkomajuri) [reply plugin-server/tests/job-queue-manager.test.ts:24]: There's tests for _enqueue error handling and error handling in the retries flow. This function just calls both of those, what do you mean?

**Comment by yakkomajuri** [comment]
Will address some more follow ups tomorrow
